### ISSUE TITLE
Skip the `isTTY` check for printing the "SUPPORTED TYPESCRIPT VERSIONS" message if the user passes in the `loggerFn` option

### ISSUE THREAD

**Issue by octogonz** [typescript-eslint/typescript-eslint#3506]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I spent 30mins tonight trying to determine why this warning was being displayed sometimes, but then it would disappear whenever I tried to debug it:

> WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

Is it a caching problem?  Is the TypeScript version actually correct when invoking ESLint programmatically?  No, turns out the magic behavior is based on `process.stdout?.isTTY`:

https://github.com/typescript-eslint/typescript-eslint/blob/44452dd5a545665c2365f2f747fc7610ea0eb686/packages/typescript-estree/src/parser.ts#L323-L330

Maybe the intent was to hide the warning from a CI build, but show it when humans invoke ESLint? That was not the effect. The effect was nondeterministic behavior that was frustrating to investigate.  Arguably this is notifying about a nontrivial problem; we don't really want our production builds linting with an outdated TypeScript parser.

Would you accept a PR to delete this `process.stdout?.isTTY` logic and always show the warning regardless of how ESLint is invoked?

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `4.26.1` |
| `TypeScript`                           | `4.3.2` |
| `node`                                 | `12.20.1` |

**Comment by bradzacher** [comment]
The intent behind this code is to support IDE tools correctly.
Certain IDE tools like vim will invoke ESLint in its `--json` mode - then read and parse the output from stdout.
In these usecases - node it is not in TTY mode.

If we indiscriminately log all the time then tools that parse JSON output will simply **hard fail** because the warning output is not valid JSON. See #1121

The `process !== undefined` check is specifically there so the parser doesn't crash when run from a browser environment (i.e. via eslint-browser, astexplorer, or prettier's browser build).

**Comment by octogonz** [comment]
> The intent behind this code is to support IDE tools correctly.
> Certain IDE tools like vim will invoke ESLint in its `--json` mode - then read and parse the JSON output from stdout.
> In these usecases - node it is not in TTY mode.

@bradzacher  That problem seems completely unrelated:  For tools that use `--json`, the warning should be reported in JSON format or else skipped entirely.

The problem that I reported affects people who invoke `eslint` from the command line without `--json`.

**Comment by bradzacher** [comment]
To correct my mistake - it's not `--json`, it's `--format json`

> For tools that use --json, the warning should be reported in JSON format or else skipped entirely

ESLint does not tell the parser what format style it's configured to use - so we do not know. The only two arguments ESLint passes to a parser are the code string, and the parser options object.

https://github.com/eslint/eslint/blob/a6a7438502abc6a1e29ec35cfbe2058ffc0803b1/lib/linter/linter.js#L659

Additionally we determine this by parsing CLI args either because if you execute ESLint via their nodejs API, you control the formatter used on the results separately to the lint run.

https://eslint.org/docs/developer-guide/nodejs-api#clienginegetformatter

I believe that's actually how these external tools generally work, and they run on the assumption that the ESLint lint run writes no console logs because ESLint itself (and its standard parsers/plugins) do not log anything.

So there is no way for us to know. Hence we have to use the fuzzy check of `isTTY`.

**Comment by octogonz** [comment]
Hmm... assuming ESLint's design is not going to change, this seems to imply that either:

- ESLint rules should not call `console.log()` in production, only when the rule itself is being debugged; OR
- `--format json` should not have been implemented because it's going to constantly get broken when rules call `console.log()`

This is interesting:
```ts
  /**
   * Allow the user to override the function used for logging
   */
  if (typeof options.loggerFn === 'function') {
    extra.log = options.loggerFn;
  } else if (options.loggerFn === false) {
    extra.log = (): void => {};
  }
```

Our toolchain invokes ESLint programmatically and could probably pass this down via the API.

Maybe passing the `loggerFn` option should disable the `isTTY` heuristic?  That would enable professional toolchains to get deterministic behavior, while still trying to guess whether `console.log()` is safe for simplistic setups that probably don't care about these nuances.

**Comment by bradzacher** [comment]
> ESLint rules should not call console.log() in production, only when the rule itself is being debugged

Correct - they should never log anything.
Ultimately - by-and-large most tooling now-a-days can correctly handle bad log lines.
But there are some tools that are just naive and make certain assumptions.

ESLint's JSON formatter was designed to allow you to easily programmatically parse ESLint's output.
It's nice because it allows you to pipe the output to another CLI tool without having to write a nodejs script.
I think most usecases today will just use the nodejs API to interact with it though.

> Our toolchain invokes ESLint programmatically and could probably pass this down via the API.

we technically will just pass whatever properties through to the underlying parser, so yeah you can pass in `loggerFn` - as long as you don't set `warnOnUnsupportedTypeScriptVersion`.

> Maybe passing the loggerFn option should disable the isTTY heuristic?

i'd be happy to accept a PR which does this

**Comment by octogonz** [comment]
> i'd be happy to accept a PR which does this

Cool.  Could we reopen this issue then?  I'll make a PR when I get some time.

**Comment by JoshuaKGoldberg** [comment]
👋 @octogonz, is this still something of interest to you?

(since it's labeled `good first issue` I don't want potential first contributors to be dissuaded by the discussion)

**Comment by octogonz** [comment]
If someone else wants to make the PR, go for it. 
  
It seems like a pretty simple fix, but myself I wouldn't work on this without also implementing the other part, where we update the Heft toolchain to wrap `loggerFn` and redirect it to the integrated logger. (Why have I not followed up sooner? I think we got better about keeping the NPM packages up to date. 😇)
